### PR TITLE
SimCluster-LayerCluster associator implementation

### PR DESCRIPTION
**Pull request by lecriste**
#### PR description:

This PR introduces an associator for HGCAL validation at SimCluster level, which will be used in a later PR. No changes are expected in the output.

#### PR validation:

`runTheMatrix -l limited`

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by lecriste** [comment]
@civanch can you please review this PR?

File: SimCalorimetry/HGCalAssociatorProducers/plugins/SimClusterAssociatorByEnergyScoreImpl.cc [hunk 1]
```diff
@@ -0,0 +1,523 @@
+#include "SimClusterAssociatorByEnergyScoreImpl.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+
+SimClusterAssociatorByEnergyScoreImpl::SimClusterAssociatorByEnergyScoreImpl(
+    edm::EDProductGetter const& productGetter,
+    bool hardScatterOnly,
+    std::shared_ptr<hgcal::RecHitTools> recHitTools,
```
Comment (civanch) [inline SimCalorimetry/HGCalAssociatorProducers/plugins/SimClusterAssociatorByEnergyScoreImpl.cc:9]: @lecriste , you really need shared_ptr here? It is shared object?

Comment (civanch) [reply SimCalorimetry/HGCalAssociatorProducers/plugins/SimClusterAssociatorByEnergyScoreImpl.cc:9]: Because the producer is global, shared_ptr is OK.

**Comment by civanch** [comment]
+1

**Comment by lecriste** [comment]
@cms-sw/upgrade-l2 can you please review this PR?

**Comment by srimanob** [comment]
+upgrade

**Comment by silviodonato** [comment]
+1
As said in the PR description, this new code will be used in a subsequent PR.
@lecriste I assume that in the next PR you are going to add some automatic test for testing this new code (eg. relvals, unit test)

**Pull request merged**